### DISCUSSION TITLE
OpenBSD ports maintainer

### DISCUSSION THREAD

**Discussion by connortechnology** [ZoneMinder/zoneminder#3948] [General]
Anyone want to take on getting ZM into open bsd ports?  I made it compile and work on openbsd, but am stuck on the actual port.

**Comment by connortechnology** [comment]
@abishai any interest?

**Comment by abishai** [comment]
Probably not, I'm not familiar with OpenBSD. It's better if a real OpenBSD user gets involved.

What about other BSDs ? :) DragonflyBSD and NetBSD

**Reply by connortechnology** [reply]
If we can find someone willing to support them.  I'm happy to handle making it compile, but can't handle the entire ports ecosystem.

**Comment by ovi-vansid** [comment]
@connortechnology  Are you still looking for someone to be an OpenBSD port maintainer?  I was looking to transition my current installation to OpenBSD, only to find that there isn't a current port or package.  

I'll be blunt, I'm not an OpenBSD expert (or Linux, for that matter), but I'm sure I could learn how to build and maintain a port.

**Comment by connortechnology** [comment]
Yes!  Duties are basically try to build from time to time and let me know of issues.  We can work together, on it.

**Comment by gdt** [comment]
Feel free to look at the pkgsrc (NetBSD) port for inspiration.  It's currently two micros behind.  I suspect there is a lot that needs to be expressed the openbsd ports way vs the pkgsrc way, but the list of things needing accomodation is likely similar.

**Comment by ovi-vansid** [comment]
Thank you, I will definitely take a look.  I've not worked on this port for
several weeks, but actually had some time this weekend.  I've got
ZoneMinder up and running on a new OpenBSD 7.8 instance, I'm just trying to
finalize documentation and configuration.  I don't really know much about
the OpenBSD porting process, so this will be a learning experience for me.
Happy to report that it actually does work, just needs to be packaged and
tested.

-vansid.

**Comment by ovi-vansid** [comment]
I had some time to work on the OpenBSD version of ZoneMinder, and I have been able to get pretty much everything working, including the API support (which was quite a pain due to missing PHP packages in the standard OpenBSD repository) and standard `rcctl` service startup. There are several issues which need to be addressed that I need some help with before it can be packaged and submitted to OpenBSD Ports, so any assistance is appreciated.  If I need to file some reports/issues or diffs let me know.

- When running ZoneMinder's `make install`, over 3,000 Perl modules get installed to the root of my OpenBSD instance.  An example is `ZoneMinder::Tag.3p`.  I'm sure they belongs elsewhere, but they get dumped in the root '/' (something to do with the ONVIF module installation?).  On my system, I had to manually copy these modules to `/usr/local/libdata/perl5/site_perl/amd64-openbsd/`. There's probably an issue somewhere in the `make` script, but I haven't looked into it yet.
- There are a couple of missing dependencies with PHP.  I had to manually download, configure, compile and install the `mysqlnd` and `apcu` PHP extensions, and there were some odd errors and problems that I encountered, such as needing to symlink files in order for the compiling to succeed.  Speaking of symlinking files...
- I don't understand what OpenBSD's equivalent to `/dev/shm` is, so I had to create it as a symlink from `/tmp`.  Ugly, but it worked.
- Also had to link a basic header file in `/usr/include/net` from `ethernet.h` to `ethertypes.h` in order for ZoneMinder to compile.  This is probably some remnant of divergence between Linux code and *BSD code (doesn't seem to be an issue in FreeBSD, either). Not sure how to address this - hopefully I can get away with making this symlink (and the `/dev/shm` one) as part of the packing porting process.
- When ZoneMinder installs, there is no mechanism to automatically add it as a recognized service, so `rcctl` does not work. I wrote a new script that follows the OpenBSD model, it seems to work. I can submit this for inclusion in the OpenBSD branch, not sure of the process.
- A minor problem with one of the Perl modules that prevents `zmpkg.pl` from starting properly, just a syntax issue with `General.pm`.  I can submit a diff file for this - would I do that as an Issue in GitHub?

I've been documenting all of my necessary changes an configurations to a fully-patched OpenBSD 7.8 installation, so I can successfully reproduce this and get ZoneMinder installed in a repeatable fashion.  But the missing PHP modules and ugly symlinks are preventing me from being able to create a proper OpenBSD port.  I have asked on the OpenBSD Ports mailing list if someone can make `mysqlnd` and `apcu` ports for PHP, as these will be critical dependencies, not heard anything back yet.  In summary, it does work using the `openbsd` branch (which is currently at 1.37.73) and my minor modifications.

**Reply by gdt** [reply]
Commenting based on experience maintaining zoneminder in pkgsrc.  Some of this is a little redundant with the previous comment (made after I typed some of it but I noticed it :-).

> * When running ZoneMinder's `make install`, over 3,000 Perl modules get installed to the root of my OpenBSD instance.  An example is `ZoneMinder::Tag.3p`.  I'm sure they belongs elsewhere, but they get dumped in the root '/' (something to do with the ONVIF module installation?).  On my system, I had to manually copy these modules to `/usr/local/libdata/perl5/site_perl/amd64-openbsd/`. There's probably an issue somewhere in the `make` script, but I haven't looked into it yet.

zm is complicated to package because it's got c++, perl, and php.  Surely openbsd has a normal way to build perl modules, and probably there's some standard idiom to include a make fragment to declare that it builds a  perl module. On NetBSD/pkgsrc, that's enough to put perl stuff in the right place.  It does sound like there is some variable that controls where they go that is ending up empty instead.

Also, try gmake if you are using BSD make.   Usually that's not necessary with cmake, but lots of makefiles use GNU extensions and fail non-obviously with BSD make.

> * There are a couple of missing dependencies with PHP.  I had to manually download, configure, compile and install the `mysqlnd` and `apcu` PHP extensions, and there were some odd errors and problems that I encountered, such as needing to symlink files in order for the compiling to succeed.  Speaking of symlinking files...

Sure, but that's not really about zoneminder packaging as needing to package those first.   However I do not recognize `mysqlnd`, it's not in pkgsrc, and I didn't notice that I needed it.

> * I don't understand what OpenBSD's equivalent to `/dev/shm` is, so I had to create it as a symlink from `/tmp`.  Ugly, but it worked.

On NetBSD,
```
# Choose directory where zoneminder can create shm files (for zmc).                                                                                                                                                                                 
# \todo Define this per-OS.                                                                                                                                                                                                                         
ZM_SHM=                 /var/shm        # NetBSD                                                                                                                                                                                                    
CMAKE_CONFIGURE_ARGS+=  -DZM_PATH_MAP=${ZM_SHM}
```
and really it's just a directory where zm can create files for shm segments.   It already exists and is 1777, and is a tmpfs.

> * Also had to link a basic header file in `/usr/include/net` from `ethernet.h` to `ethertypes.h` in order for ZoneMinder to compile.  This is probably some remnant of divergence between Linux code and *BSD code (doesn't seem to be an issue in FreeBSD, either). Not sure how to address this - hopefully I can get away with making this symlink (and the `/dev/shm` one) as part of the packing porting process.

I didn't have to do this, and NetBSD also lacks ethernet.h and has ethertypes.h.  (Most  of this arose in 4.2BSD to 4.4BSD or earlier, and Linux reinvented a bunch of it differently.)

> * When ZoneMinder installs, there is no mechanism to automatically add it as a recognized service, so `rcctl` does not work. I wrote a new script that follows the OpenBSD model, it seems to work. I can submit this for inclusion in the OpenBSD branch, not sure of the process.

Sometimes things like that go in upstream, and sometimes they are just files in the package.
 
> * A minor problem with one of the Perl modules that prevents `zmpkg.pl` from starting properly, just a syntax issue with `General.pm`.  I can submit a diff file for this - would I do that as an Issue in GitHub?

Sounds like a PR makes sense.  (With any luck the openbsd branch can be merged to master soon.)

**Reply by ovi-vansid** [reply]
Really good info, I appreciate it. I've noticed a couple of people have mentioned (such as @abishai  in a note below) that `mysqlnd` is not needed...but I definitely got errors connecting PHP to the backend database until I compiled and installed this module.  Is there some alternate configuration elsewhere in ZM that I could change so I don't need this and can use `pdo_mysql` instead?

**Reply by ovi-vansid** [reply]
> Sounds like a PR makes sense. (With any luck the openbsd branch can be merged to master soon.)

Yep, it was merged this week.  I just submitted [a PR](https://github.com/ZoneMinder/zoneminder/pull/5007).

**Comment by abishai** [comment]
> over 3,000 Perl modules get installed to the root of my OpenBSD instance

Probably the solution is to tweak ZM_PERL_MM_PARMS variable in CMake

> what OpenBSD's equivalent to /dev/shm

Any tmpfs directory. Can be set in CMake ZM_PATH_MAP or configured later. Setting this directory not in tmpfs backed filesystem would be a disaster.

> mysqlnd and apcu ports for PHP, as these will be critical dependencies

They are not. You can throw away apcu entirely and use pdo_mysql instead of mysqlnd

**Reply by ovi-vansid** [reply]
> You can throw away apcu entirely 

I think you're right for the general usage of ZoneMinder, but in my testing I could not get the API to work without it.  Is this a correct interpretation on my part?  I'd have to rebuild my test environment to capture the error logs again, I just won't have time for a week or so.

> use pdo_mysql instead of mysqlnd

I also asked about this in a reply above - how do I use that module instead?  Sorry for such a rookie question, but it isn't clear where I can swap-out modules and my testing seemed to indicate that ZM would not connect to the database without it.

**Comment by abishai** [comment]
> I could not get the API to work 

I believe you are right. apc(u) is really required. sysvsem is recommended as well (workaround tries to write in script directory, which is very bad idea).

> how do I use that module instead

pdo is middleware, it can link against mysqlnd or mysql-client with make flags. The application should be unaware of that, because it works using pdo abstraction. Probably, your issues is not coming from pdo. Do you have logs?

**Comment by gdt** [comment]
I have `php-pdo` and `php-pdo_mysql`.  I also have `php-apchu`.

My impression is that zm used to use sysv shm, but now uses mmap.    But you said `sysvsem`.  I am not doing anything intentional about sysv semaphores.

**Comment by connortechnology** [comment]
ZM still has the code to use sysv shm, but no one uses it anymore, but there's not reason it shouldn't still work.
sysvsem is a php package thing for preventing concurrent access to sockets when talking to zms.   Everything should continue to work without it, but you get warnings in  the logs without it.

**Comment by gdt** [comment]
Are you saying about shm:

   there is code to use sysv shm if configured (instead of mmap).  Even though it's ~never used by anyone, a package that lacks it is defective?    or "it's never used, and if you install php-sysvshm it should work, but that's borderline historical and it's fine for a package to omit it.

And about sysvsem:

  It is always used, to serialize socket access for zms.  If it's missing, then 1) there will be warnings logged and 2) there will be no seralization control, so you are left just hoping that there aren't concurrent accesses

**Comment by abishai** [comment]
I believe. there is script, that queries zms periodically for monitor data, like fps. It will fail without sysvsem php package. Or if fallback fails to create semaphore file (because it trites to create in script location directory - it should be read only).

**Comment by connortechnology** [comment]
sysvshm vs mmap is a compile time flag.  You need one of them.  Either/or.  It is not a run time flag.  

php-sysvsem is not REQUIRED, it is used to serialize socket communication to zms.  No one ever saw a problem, but there was potential.  It will log a warning, and then continue.

**Comment by ovi-vansid** [comment]
Sorry for the long delay, I ended up moving and starting a new job, took a lot longer than expected to get settled and start work on this port again.  I dusted off my VMs, documentation, etc. and built a working ZoneMinder running on the latest OpenBSD 7.9 (x64).  Unfortunately, it looks like the `openbsd` branch is still v1.37.73 - @connortechnology , is this right? Trying to build this from the `master` branch doesn't seem like the right way to go, but just wanted to check to see if I did something wrong (git is not my strong suit).

**Comment by connortechnology** [comment]
I can update it to master.  SHould be a simple merge.

**Comment by connortechnology** [comment]
merge done

**Comment by ovi-vansid** [comment]
Just a quick update on my progress: 

I'm about 95% of the way done with an OpenBSD port using the [ports tree Makefile process](https://www.openbsd.org/faq/ports/guide.html) that should be ready for testing soon.  Running into a small problem right at the end trying to figure out why I'm getting errors related to PLIST files so I'll post something on the openbsd-ports mailing list for some help.  

I've otherwise got ZoneMinder 1.39.17 to compile and install properly on OpenBSD 7.9, so it's just a matter of getting the port/Makefile/etc. completed. There are probably a few tweaks that I'll submit as PRs so that I'm not trying to package a customized version that will be harder to maintain.

**Reply by ovi-vansid** [reply]
99% of the way there, but I've run into a very small problem (nothing that `sed` can't handle, but that's probably not the best approach in the long term). 

I submitted a [PR to update the CMakeLists.txt](https://github.com/ZoneMinder/zoneminder/pull/5045) file to handle certain OpenBSD cases, but it doesn't seem to be respected.  I should also be able to have an entry in the OpenBSD port Makefile `CMAKE_ARGS +=   -DZM_PATH_MAP:PATH=/tmp` , but that also seems to be ignored.  Like I said, I'm working around it with `sed` changing the text in one of the `zm.conf` files for right now, but @connortechnology do you think there's an issue with the distro logic in `CMakeLists.txt`?

**Comment by gdt** [comment]
I will try to look more later, but I believe that in pkgsrc the CMAKE_CONFIGURE_ARGS settings to set paths are mostly working.
If I can set using cmake args, I have not worried about defaults.

**Comment by connortechnology** [comment]
Hey, just did an update on my box here.. had issues.  Our default cakephp cache is set to APc which isn't installed, we need to set    -DZM_CAKEPHP_CACHE=File when running cmake.  ANd it wanted to put cache files in /var/cache/zoneminder/tempcache which I had to create as well as  /var/cache/zoneminder/tempcache/models and /var/cache/zoneminder/tempcache/persistent and chown them.  So that needs to be figured out and improved.

**Reply by ovi-vansid** [reply]
@gdt  I've tried `CMAKE_CONFIGURE_ARGS += -DZM_PATH_MAP:PATH=/tmp` and that didn't seem to work.  Still ended up with `ZM_PATH_MAP=/dev/shm` in `/usr/local/etc/zm/conf.d/01-system-paths.conf`.  Like I said, I can work around it using an `@exec` statement in `pkg/PLIST` that runs post-installation.  

Speaking of that, I also "fixed" the directory issues @connortechnology mentioned above the same way, I just have another `@exec` statement that creates the directories and sets the correct permissions.  I'm sure we can fix it the right way in the main codebase.

Also, I have included php/apcu as a dependency in the ports Makefile, seems to work fine.  Let me attach what I have so far here if you want to test it on an OpenBSD 7.9 system (not sure how well it will work on older installations, never tested).  

This is a "works for me" draft version.  You'll need to:

1. Setup ports on your OpenBSD installation
```
cd /tmp
ftp https://cdn.openbsd.org/pub/OpenBSD/$(uname -r)/{ports.tar.gz,SHA256.sig}
signify -Cp /etc/signify/openbsd-$(uname -r | cut -c 1,3)-base.pub -x SHA256.sig ports.tar.gz
cd /usr
doas tar xzf /tmp/ports.tar.gz
```
2. Copy `zoneminder-1.39.17-openbsd-port.tar.gz` to `/usr/ports/`
```
doas cp /path/to/zoneminder-1.39.17-openbsd-port.tar.gz /usr/ports/
```

3. Extract the files from within `/usr/ports/`, which will create a `mystuff/security/zoneminder` structure
```
doas tar xvf zoneminder-1.39.17-openbsd-port.tar.gz
```

4. From the `/usr/ports/mystuff/security/zoneminder` directory, install the port using these steps
```
make makesum
make configure
make build
make update-plist
make install
```

- **_NOTE_**: If you don't have `FETCH_PACKAGES=` (yes, nothing after the equal) in `/etc/mk.conf`, the `make configure` step **WILL TAKE HOURS, POSSIBLY DAYS** (depending on your connection and system speed) as every single dependent package will need to be downloaded and compiled from source.  When you reach the point that you're compiling multiple compilers (rust, llvm, etc.) you'll notice why this was probably not a good idea.
- For the uninitiated (like me), there are some intricacies with the OpenBSD make process, so that will be an exercise for the reader.   Just know that you might need to run some of the Step 4 commands as root/doas, or know how to configure `/etc/mk.conf` for privilege separation. 

5. There are a bunch of post-install steps that come with pretty much every ZoneMinder installation, like setting up the database, configuring Apache, etc.  I cover those in a file from `pkg/MESSAGE` that instructions the user after the final package is installed.

There are a couple of ugly workarounds, but it pretty much works, including the API.  Please give me some feedback and let me know your collective thoughts before I post it to the openbsd-ports mailing list.

[zoneminder-1.39.17-openbsd-port.tar.gz](https://github.com/user-attachments/files/31115472/zoneminder-1.39.17-openbsd-port.tar.gz)

**UPDATE**: The correct syntax is `CONFIGURE_ARGS += -DZM_PATH_MAP:PATH=/tmp`, seems to be working.  I'm going to post a small tarball to openbsd-ports and see where it goes.  This is certainly a first draft, so more changes will need to be made.

**Comment by gdt** [comment]
I don't understand how openbsd ports spells cmake command args, and didn't mean to imply it was or should be the same as pkgsrc.   Just that I think I'm successfully setting them.  I encourage you to trace the build and look at the cmake cache, because there are multiple steps through to getting the example config file adjusted.
